### PR TITLE
New tests for group constructors and some fixes

### PR DESCRIPTION
**Pull request by fingolfin**
* add a bunch of tests for group constructors in the grp directory
* add tests for for various matrix groups that verify the given invariant forms
* fix GO(1,4,5) (triggers an error on master)
* add invariant forms for Sp(2*n, Integers mod 2^k)
* remove cache limit from perfect groups library, to increase speed
* remove some invalid (and unusable) constructor variants
* tweak error messages in some group constructors

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
These all look good and uncontroversial. Will merge unless someone says something in the next couple of days.

It is annoying it looks like coverage got worse, as tst files aren't marked as covered I think.

**Comment by fingolfin** [comment]
Ahhhh, is that why? hmmm, so how are the total lines reported to / computed by codecov? I would havethought it only counts source files in lib, src, ... as shown on their website...

**Comment by fingolfin** [comment]
I think the tes files are not the reason for this change. Rather, what really happened is that codecov ignores files which have neither hits nor misses. With this PR, it soundly "learned" about some additional source files (several `perf*.grp` files) which are not covered.

**Comment by fingolfin** [comment]
Something weird is going on: I added a brutal test which invokes `PerfectGroup` for every legal input. So it should in theory show 100% coverage for all the `perfNN.grp` files. Yet it doesn't, it even reports 0% coverage for some of them. Huh?

@ChrisJefferson any idea what might be going wrong here?

**Comment by ChrisJefferson** [comment]
Good question -- not sure.

If I manually run `PerfectGroup(7800)`, then I get coverage of the first part of `perf2.grp`, so it's some combination of the tests being run, and the information going to testcov I think. For some reason, I'm having trouble checking out your branch for some reason to try it out.

**Comment by ChrisJefferson** [comment]
Nope, it is my fault, something weird is going on (your tests lead to weirdly corrupted profiling output. I think it's because the files keep getting re-read. Investigating).

**Comment by ChrisJefferson** [comment]
Turns out profiling breaks with re-reading of files! Will looking into fixing that.

Independently, the Perfect Groups library only keeps a cache of 50 groups (if I'm reading it right), perf.grp, line 31. If I push that to 1000, then `perf.grp` goes from 2.7 sec to 0.5 sec.

**Comment by fingolfin** [comment]
Nice catch about that catch. In fact, there are only 331 entries, so setting this to 1000 just caches everything.

I just checked: a fully loaded `PERFGRP` is about 4.6 MB (though this is without my `MemoryUsage` fixes...). I think we should just remove the caching logic. And perhaps in the future revise it, once we got around to adding a proper generic "cache" system (possibly using weak pointers and stuff....) which we need for the HPC-GAP integration anyway.

**Comment by ChrisJefferson** [comment]
Coverage looks much better now -- no idea why one tiny bit of perf1.grp isn't covered... could you have a look and see if you think it's a profiling problem, or a GAP problem?

**Comment by fingolfin** [comment]
The unused code in `perf1.grp` really is not called, and there is a comment next to it which says `  # 1920.8 (otherpres.)`. I will investigate (just got the book by Plesken & Holt on my desktop)

**Comment by fingolfin** [comment]
Actually, there are a bunch of these entries with an "otherpres." comment next to them. The indices of these entries are always higher than the number of perfect groups of that order, so presumably we are looking at alternative presentations for some groups (but which?). So the *real* weird part is that some of them *are* used after all... indirectly. E.g. 1080.2 is used by `PerfectGroup(69120,1)`.

I am not sure why 1920.8, 1344.3 and 1344.4 are there but not used. I could not yet identify their source in Holt/Plesken (but I only skimmed and didn't try hard, so it might be in there). Anyway, it seems harmless.

**Comment by fingolfin** [comment]
I think this is good now. Overall coverage is up by 3.55% and the `grp` dir has 90% coverage :-). Of course this makes it look at bit better than it is: There certainly could be more comprhensive checks. E.g. one could check whether the pre-computed group sizes and other properties are correct (by doing something like `H:=Group(GeneratorsOfGroup(G));; Size(G)=Size(H);`), check other group properties, etc. But it's a start.

File: grp/classic.gi [hunk 1]
```diff
@@ -403,7 +403,7 @@ BindGlobal( "Oplus45", function()
 
     # construct the group without calling 'Group'
     g := [ phi*tau2, tau*eichler*delta ];
-    g:=List(g,i->ImmutableMatrix(f,i),true);
+    g:=List(g,i->ImmutableMatrix(f,i));
```
Comment (ChrisJefferson) [inline grp/classic.gi]: Did this code ever work, out of interest? What's that `true` supposed to do?

File: grp/glzmodmz.gi [hunk 2]
```diff
@@ -170,7 +170,10 @@ local t,geni,m,slmats,gens,f,rels,i,j,k,l,mat,mat1,mats,id,nh;
     od;
   od;
 
-  return Group(slmats);
+  g := Group(slmats);
+  mat := Concatenation(id{[nh+1..n]},-id{[1..nh]});
+  SetInvariantBilinearForm(g,rec(matrix:=mat));
+  return g;
 end);
```
Comment (ChrisJefferson) [inline grp/glzmodmz.gi]: I'm going to trust this is right.. I assume these is a test for it?

File: grp/classic.gi (hunk 1 above)
Comment (fingolfin) [inline grp/classic.gi]: Yes, it worked until 814a0f6e77bbd0472619a588188c1858d5e4e6e3 broke it by making `List` validate its arguments more agressively. My guess is that the `true` in there was an accident.

File: grp/glzmodmz.gi (hunk 2 above)
Comment (fingolfin) [inline grp/glzmodmz.gi]: Yes -- indeed, writing the tests for `Sp(2*n, Integers mod p^k)` lead me to discover that the invariant form was missing for `p=2`, and hence I added this to make the tests pass. so this was TDD (test driven development) :-).

**Comment by fingolfin** [comment]
So, merge this? Rebase it? Should I do it myself?

**Pull request merged**